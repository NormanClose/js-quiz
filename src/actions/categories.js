import database from '../firebase/firebase';

// ADD_CATEGORY
export const addCategory = (category) => ({
  type: 'ADD_CATEGORY',
  category
});

export const startAddCategory = (categoryData = {}) => {
  return (dispatch, getState) => {

    const {
      category = ''
    } = categoryData;
    const categoryValue = { category };
console.log('categoryValue', categoryValue);
    return database.ref(`/categories`).push(categoryValue).then((ref) => {
      dispatch(addCategory({
        id: ref.key,
        ...categoryValue
      }));
    });
  };
};

// SET_CATEGORIES
export const setCategories = (categories) => ({
  type: 'SET_CATEGORIES',
  categories
});

export const startSetCategories = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    return database.ref(`/categories`).once('value').then((snapshot) => {
      const categories = [];

      snapshot.forEach((childSnapshot) => {
        categories.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        });
      });
      dispatch(setCategories(categories));
    });
  };
};