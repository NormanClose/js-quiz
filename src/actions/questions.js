//import uuid from 'uuid';
import database from '../firebase/firebase';

// ADD_QUESTION
export const addQuestion = (question) => ({
  type: 'ADD_QUESTION',
  question
});

export const startAddQuestion = (questionData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    const {
      language = '',
      level = 1,
      category,
      question = '',
      questionType = 'single',
      answer1Correct = false,
      answer1 = '',
      answer2Correct = false,
      answer2 = '',
      answer3Correct = false,
      answer3 = '',
      answer4Correct = false,
      answer4 = '',
      answer5Correct = false,
      answer5 = '',
      answer6Correct = false,
      answer6 = ''
    } = questionData;
    const questionInput = { language,level,category,question,questionType,answer1Correct,answer1,answer2Correct,answer2,answer3Correct,answer3,answer4Correct,answer4,answer5Correct,answer5,answer6Correct,answer6};

console.log('user is ' + uid);
//does not work - ".write": "data.parent().child('admins').child(auth.uid).exists()"

    return database.ref(`/questions`).push(questionInput).then((ref) => {
      dispatch(addQuestion({
        id: ref.key,
        ...questionInput
      }));
    });
  };
};

// REMOVE_QUESTION
export const removeQuestion = ({ id } = {}) => ({
  type: 'REMOVE_QUESTION',
  id
});

export const startRemoveQuestion = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    return database.ref(`/questions/${id}`).remove().then(() => {
      dispatch(removeQuestion({ id }));
    });
  };
};

// EDIT_QUESTION
export const editQuestion = (id, updates) => ({
  type: 'EDIT_QUESTION',
  id,
  updates
});

export const startEditQuestion = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    return database.ref(`/questions/${id}`).update(updates).then(() => {
      dispatch(editQuestion (id, updates))
    })
  }
};

// SET_QUESTIONS
export const setQuestions = (questions) => ({
  type: 'SET_QUESTIONS',
  questions
});

export const startSetQuestions = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    return database.ref(`/questions`).once('value').then((snapshot) => {
      const questions = [];

      snapshot.forEach((childSnapshot) => {
        questions.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        });
      });

      dispatch(setQuestions(questions));
    });
  };
};
