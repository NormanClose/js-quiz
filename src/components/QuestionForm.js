import React from 'react';
import CategoriesFormSelectBox from './CategoriesFormSelectBox';
import { withFormik, Form, Field } from 'formik';
import * as Yup from 'yup'; //Note change with V0.25

const InnerForm= ({ values, errors, touched, isSubmitting, setFieldValue, handleChange, onCancel}) => (
  <div>
    <Form className='form'>
      <div>
        <div>
          <button disabled={isSubmitting} className='button button--form'>
          Submit
          </button>
        </div>
        <div className='spacer--med'>
        </div>
        <div>
          <label className='label'>
            Language 
            <Field component='select' name='language' className='select'>
              <option value='javascript'>Javascript</option>
            </Field>
          </label>
        </div>
        <div className='radio-group'>
          <label className='tick-container'>
            Level 1
            <Field type='radio' name='level' id='1' value='1' checked={values.level ==='1'} />
            <span className='radio-checkmark'></span>
          </label>
          <label className='tick-container'>
            Level 2
            <Field type='radio' name='level' id='2' value='2' checked={values.level ==='2'} />
            <span className='radio-checkmark'></span>
          </label>
          <label className='tick-container'>
            Level 3
            <Field type='radio' name='level' id='3' value='3' checked={values.level ==='3'} />
            <span className='radio-checkmark'></span>
          </label>
        </div>
        <div>
          <label className='label'>
            Category
            <CategoriesFormSelectBox 
              categories={values.categories} 
              onChange={setFieldValue}
              value={values.category}
            />
          </label>
        </div>
        <p className='error'>{errors.question}</p>
        <Field
          component = 'textarea'
          className = 'textarea'
          name='question'
          placeholder='Question'
        />
        <div>
          <label className='label'>
            Question Type 
            <Field component='select' name='questionType' className='select'>
              <option value='single'>Only one correct answer</option>
              <option value='multi'>More than one correct</option>
            </Field>
          </label>
        </div>
        <p className='error'>{errors.overall}</p>
        <div>
          <p>{errors.answer1Correct}</p>
          <label className='tick-container'>
            Answer 1 Correct
            <Field type='checkbox' name='answer1Correct' checked={values.answer1Correct === true}/>
            <span className='box-checkmark'></span>
          </label>
        </div>
        <p className='error'>{errors.answer1}</p>
        <Field
          component = 'textarea'
          className = 'textarea'
          name='answer1'
          placeholder='Answer 1'
        />
        <div>
          <label className='tick-container'>
            Answer 2 Correct
            <Field type='checkbox' name='answer2Correct' checked={values.answer2Correct === true}/>
            <span className='box-checkmark'></span>
          </label>
        </div>
        <p className='error'>{errors.answer2Correct}</p>
        <Field
          component = 'textarea'
          className = 'textarea'
          name='answer2'
          placeholder='Answer 2'
        />
        <div>
          <label className='tick-container'>
            Answer 3 Correct
            <Field type='checkbox' name='answer3Correct' checked={values.answer3Correct === true}/>
            <span className='box-checkmark'></span>
          </label>
        </div>
        <p className='error'>{errors.answer3Correct}</p>
        <Field
          component = 'textarea'
          className = 'textarea'
          name='answer3'
          placeholder='Answer 3'
        />
        <div>
          <label className='tick-container'>
            Answer 4 Correct
            <Field type='checkbox' name='answer4Correct' checked={values.answer4Correct === true}/>
            <span className='box-checkmark'></span>
          </label>
        </div>
        <p className='error'>{errors.answer4Correct}</p>
        <Field
          component = 'textarea'
          className = 'textarea'
          name='answer4'
          placeholder='Answer 4'
        />
        <div>
          <label className='tick-container'>
            Answer 5 Correct
            <Field type='checkbox' name='answer5Correct' checked={values.answer5Correct === true}/>
            <span className='box-checkmark'></span>
          </label>
        </div>
        <p className='error'>{errors.answer5Correct}</p>
        <Field
          component = 'textarea'
          className = 'textarea'
          name='answer5'
          placeholder='Answer 5'
        />
        <div>
          <label className='tick-container'>
            Answer 6 Correct
            <Field type='checkbox' name='answer6Correct' checked={values.answer6Correct === true}/>
            <span className='box-checkmark'></span>
          </label>
        </div>
        <p className='error'>{errors.answer6Correct}</p>
        <Field
          component = 'textarea'
          className = 'textarea'
          name='answer6'
          placeholder='Answer 6'
        />
      </div>
      <button type='submit' disabled={isSubmitting} className='button button--form'>
        Submit
      </button>
      <button type='button' disabled={isSubmitting} onClick={onCancel}>
        Cancel
      </button>
    </Form>
  </div>
)

const PostForm = withFormik({
  mapPropsToValues: props => {
    return {
      language : props.question.language || '',
      level : props.question.level || '1',
      category : props.question.category || '',
      question : props.question.question || '',
      questionType : props.question.questionType || 'single',
      answer1Correct : props.question.answer1Correct || false,
      answer1 : props.question.answer1 || '',
      answer2Correct : props.question.answer2Correct || false,
      answer2 : props.question.answer2 || '',
      answer3Correct : props.question.answer3Correct || false,
      answer3 : props.question.answer3 || '',
      answer4Correct : props.question.answer4Correct || false,
      answer4 : props.question.answer4 || '',
      answer5Correct : props.question.answer5Correct || false,
      answer5 : props.question.answer5 || '',
      answer6Correct : props.question.answer6Correct || false,
      answer6 : props.question.answer6 || '',
      categories : props.categories
    };
  },
  validationSchema: Yup.object().shape({
    question: Yup.string().min(5, 'Question must be 5 characters or longer').required('Question is required'),
    answer1: Yup.string().min(5, 'Answer 1 must be 5 characters or longer').required('Answer 1 is required')
  }),
  handleCancel: () => {
    props.onCancel();
  },
  handleSubmit: (values, { resetForm, setErrors, setSubmitting, props }) => {

    let countQuestionsCorrect = 0;
    if(values.answer1Correct) {
      countQuestionsCorrect++;}
    if(values.answer2Correct) {
      countQuestionsCorrect++;}
    if(values.answer3Correct) {
      countQuestionsCorrect++;}
    if(values.answer4Correct) {
      countQuestionsCorrect++;}
    if(values.answer5Correct) {
      countQuestionsCorrect++;}
    if(values.answer6Correct) {
      countQuestionsCorrect++;}

    let countErrors = 0;

    if (values.answer1Correct && !values.answer1) {
      countErrors++;
      setErrors({answer1Correct: 'If answer is correct then need answer filled in' });
    }
    if (values.answer2Correct && !values.answer2) {
      countErrors++;
      setErrors({answer2Correct: 'If answer is correct then need answer filled in' });
    }
    if (values.answer3Correct && !values.answer3) {
      countErrors++;
      setErrors({answer3Correct: 'If answer is correct then need answer filled in' });
    }
    if (values.answer4Correct && !values.answer4) {
      countErrors++;
      setErrors({answer4Correct: 'If answer is correct then need answer filled in' });
    }
    if (values.answer5Correct && !values.answer5) {
      countErrors++;
      setErrors({answer5Correct: 'If answer is correct then need answer filled in' });
    }
    if (values.answer6Correct && !values.answer6) {
      countErrors++;
      setErrors({answer6Correct: 'If answer is correct then need answer filled in' });
    }

    if (values.questionType === 'single') {
      if(countQuestionsCorrect === 0) {
        countErrors++;
        setErrors({overall: 'At least one answer must be marked as correct' });
      } else if(countQuestionsCorrect > 1) {
        countErrors++;
        setErrors({overall : 'Only one answer can be correct'});
      }
    } else if (values.questionType === 'multi') {
      if(countQuestionsCorrect < 2) {
        countErrors++;
        setErrors({overall: 'You have selected \'More than one correct\' - Mark more than one answer as correct' });
      }
    }

    if (countErrors === 0) {
      resetForm();
    } else {
      setSubmitting(false);
      return;
    }

    props.onSubmit(values);
  }
})(InnerForm);

export default PostForm