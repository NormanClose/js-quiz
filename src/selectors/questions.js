// Get visible questions with filters applied

export default (questions, { level, category }) => {
  return questions.filter((question) => {
    const levelMatch = question.level.toLowerCase().includes(level.toLowerCase())
    const categoryMatch = question.category.toLowerCase().includes(category.toLowerCase());

    return categoryMatch && levelMatch;
  })
};