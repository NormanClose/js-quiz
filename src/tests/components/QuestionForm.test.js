import React from 'react';
import { shallow } from 'enzyme';
import moment from 'moment';
import QuestionForm from '../../components/QuestionForm';
import questions from '../fixtures/questions';
import categories from '../fixtures/categories'

test('should render QuestionForm correctly', () => {
  const wrapper = shallow(<QuestionForm question={{}} onSubmit={{}} categories = {categories} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render QuestionForm correctly with question data', () => {
  const wrapper = shallow(<QuestionForm question={questions[1]} onSubmit={{}} categories = {categories} />);
  expect(wrapper).toMatchSnapshot();
});


