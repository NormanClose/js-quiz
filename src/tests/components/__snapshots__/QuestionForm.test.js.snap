// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render QuestionForm correctly 1`] = `
<Formik
  categories={
    Array [
      Object {
        "category": "Value",
        "id": "1",
      },
      Object {
        "category": "Reference",
        "id": "2",
      },
      Object {
        "category": "Function",
        "id": "3",
      },
    ]
  }
  enableReinitialize={false}
  handleSubmit={[Function]}
  initialValues={
    Object {
      "answer1": "",
      "answer1Correct": false,
      "answer2": "",
      "answer2Correct": false,
      "answer3": "",
      "answer3Correct": false,
      "answer4": "",
      "answer4Correct": false,
      "answer5": "",
      "answer5Correct": false,
      "answer6": "",
      "answer6Correct": false,
      "categories": Array [
        Object {
          "category": "Value",
          "id": "1",
        },
        Object {
          "category": "Reference",
          "id": "2",
        },
        Object {
          "category": "Function",
          "id": "3",
        },
      ],
      "category": "",
      "language": "",
      "level": "1",
      "question": "",
      "questionType": "single",
    }
  }
  isInitialValid={false}
  onSubmit={[Function]}
  question={Object {}}
  render={[Function]}
  validateOnBlur={true}
  validateOnChange={true}
  validationSchema={[Function]}
/>
`;

exports[`should render QuestionForm correctly with question data 1`] = `
<Formik
  categories={
    Array [
      Object {
        "category": "Value",
        "id": "1",
      },
      Object {
        "category": "Reference",
        "id": "2",
      },
      Object {
        "category": "Function",
        "id": "3",
      },
    ]
  }
  enableReinitialize={false}
  handleSubmit={[Function]}
  initialValues={
    Object {
      "answer1": "Variables are good",
      "answer1Correct": true,
      "answer2": "",
      "answer2Correct": false,
      "answer3": "",
      "answer3Correct": false,
      "answer4": "",
      "answer4Correct": false,
      "answer5": "",
      "answer5Correct": false,
      "answer6": "",
      "answer6Correct": false,
      "categories": Array [
        Object {
          "category": "Value",
          "id": "1",
        },
        Object {
          "category": "Reference",
          "id": "2",
        },
        Object {
          "category": "Function",
          "id": "3",
        },
      ],
      "category": "Variables",
      "language": "",
      "level": "3",
      "question": "Test Variables",
      "questionType": "single",
    }
  }
  isInitialValid={false}
  onSubmit={[Function]}
  question={
    Object {
      "answer1": "Variables are good",
      "answer1Correct": true,
      "answer2": "",
      "answer2Correct": false,
      "answer3": "",
      "answer3Correct": false,
      "answer4": "",
      "answer4Correct": false,
      "answer5": "",
      "answer5Correct": false,
      "answer6": "",
      "answer6Correct": false,
      "category": "Variables",
      "language": "",
      "level": "3",
      "question": "Test Variables",
      "questionType": "single",
    }
  }
  render={[Function]}
  validateOnBlur={true}
  validateOnChange={true}
  validationSchema={[Function]}
/>
`;
